policy-controller/k8s/index/src/pod.rs

```diff
@@ -0,0 +1,367 @@
+        for pod in self.index.values_mut() {
+            pod.link_servers(servers)
+        }
+    }
+
+    pub(crate) fn reset_server(&mut self, name: &str) {
+        for (pod_name, pod) in self.index.iter_mut() {
+            let rx = pod.default_allow_rx.clone();
+            for (p, port) in pod.ports.by_port.iter_mut() {
+                if port
+                    .server_name
+                    .as_ref()
+                    .map(|n| n == name)
+                    .unwrap_or(false)
+                {
+                    debug!(pod = %pod_name, port = %p, "Removing server from pod");
+                    port.server_name = None;
+                    port.server_tx
+                        .send(rx.clone())
+                        .expect("pod config receiver must still be held");
+                } else {
+                    trace!(pod = %pod_name, port = %p, server = ?port.server_name, "Server does not match");
+                }
+            }
+        }
+    }
+}
+
+// === impl Pod ===
+
+impl Pod {
+    /// Links this pods to server (by label selector).
+    //
+    // XXX This doesn't properly reset a policy when a server is removed or de-selects a pod.
+    fn link_servers(&mut self, servers: &SrvIndex) {
+        let mut remaining_ports = self.ports.by_port.keys().copied().collect::<HashSet<u16>>();
+
+        // Get all servers that match this pod.
+        let matching = servers.iter_matching(self.labels.clone());
+        for (name, port_match, rx) in matching {
+            // Get all pod ports that match this server.
+            for p in self.ports.collect_port(port_match).into_iter().flatten() {
+                self.link_server_port(p, name, rx);
+                remaining_ports.remove(&p);
+            }
+        }
+
+        // Iterate through the ports that have not been matched to clear them.
+        for p in remaining_ports.into_iter() {
+            let port = self.ports.by_port.get_mut(&p).unwrap();
+            port.server_name = None;
+            port.server_tx
+                .send(self.default_allow_rx.clone())
+                .expect("pod config receiver must still be held");
+        }
+    }
+
+    fn link_server_port(&mut self, port: u16, name: &str, rx: &ServerRx) {
+        let port = match self.ports.by_port.get_mut(&port) {
+            Some(p) => p,
+            None => return,
+        };
+
+        // Either this port is using a default allow policy, and the server name is unset, or
+        // multiple servers select this pod. If there's a conflict, we panic if the controller is
+        // running in debug mode. In release mode, we log a warning and ignore the conflicting
+        // server.
+        //
+        // TODO these cases should be prevented with a validating admission controller.
+        if let Some(sn) = port.server_name.as_ref() {
+            if sn != name {
+                debug_assert!(false, "Pod port must not match multiple servers");
```

Comment:
very nitpicky: should this be moved _after_ the `warn!` macro (which actually prints the conflicting matches), so that if the assertion fails a test, we get the debugging information out? nbd.